source/server/http/stats_handler.cc

```diff
@@ -190,52 +190,110 @@ uint64_t PrometheusStatsFormatter::statsAsPrometheus(
     const std::vector<Stats::GaugeSharedPtr>& gauges,
     const std::vector<Stats::ParentHistogramSharedPtr>& histograms, Buffer::Instance& response,
     const bool used_only, const absl::optional<std::regex>& regex) {
-  std::unordered_set<std::string> metric_type_tracker;
-  for (const auto& counter : counters) {
-    if (!shouldShowMetric(*counter, used_only, regex)) {
-      continue;
-    }
 
-    const std::string tags = formattedTags(counter->tags());
-    const std::string metric_name = metricName(counter->tagExtractedName());
-    if (metric_type_tracker.find(metric_name) == metric_type_tracker.end()) {
-      metric_type_tracker.insert(metric_name);
-      response.add(fmt::format("# TYPE {0} counter\n", metric_name));
+  /*
+   * From
+   * https:*github.com/prometheus/docs/blob/master/content/docs/instrumenting/exposition_formats.md#grouping-and-sorting:
+   *
+   * All lines for a given metric must be provided as one single group, with the optional HELP and
+   * TYPE lines first (in no particular order). Beyond that, reproducible sorting in repeated
+   * expositions is preferred but not required, i.e. do not sort if the computational cost is
+   * prohibitive.
+   */
+
+  /**
+   * Processes a metric type (counter, gauge, histogram) by generating all output lines, sorting
+   * them by tag-extracted metric name, and then outputting them in the correct sorted order into
+   * response.
+   *
+   * @param metrics A vector of Stats::RefcountPtr to a metric type.
+   * @param generate_output A std::function<std::string(const MetricType& metric, const std::string&
+   *        prefixedTagExtractedName)> which returns the output text for this metric.
+   */
+  auto process_type = [&response, &regex, used_only](const auto& metrics,
```

Comment:
I was wondering whether you could use templates in some way to reduce the user of `auto` -- the `auto` for the generated function is OK but it'd be nicer to be explicit about what type metrics and generate_output were.

See https://google.github.io/styleguide/cppguide.html#Type_deduction for discussion.

I think my concrete suggestion is to break up most of this code into a private helper method `processType` which is templatized. Then you could have the body of statsAsPrometheus() simply be:
```
  uint64_t metric_name_count = 0;
  metric_name_count += rocessType<Stats::Counter>(counters, generate_counter_and_gauge_output, "counter");
  metric_name_count += processType<Stats::Gauge>(gauges, generate_counter_and_gauge_output, "gauge");
  metric_name_count += processType<Stats::Histogram>(histograms, generate_histogram_output, "histogram");
```

I think that might be clearer than using the lambda with the auto...WDYT?

That would also remove the need for the `element_type` in the refcount, at least for this PR.

Reply:
Both `metrics` and `generate_output` are templatized types because of how `Counter`, `Gauge`, and `Histogram` are defined. There isn't a common/inherited defintion for the functions that get the data. I had tried to make these non-auto, and it isn't possible.

I can make the helpers into template methods instead of a lambda, and I agree that may make this a little easier to follow. Some of the types will become more concrete.

Reply:
I think once you make the helper into an explicit template method you'll be able to drop the `auto`, though I could be wrong.

Reply:
Yeah, it came out better than I expected, although I had to explicitly specify template parameters in some cases that were unexpected to me. But I think it is more readable.